File: core/secrets/SecretProvider.ts
import crypto from 'crypto';

export type SecretKey = string;

export interface SecretResult {
  key: SecretKey;
  value: string;
  source: SecretSource;
  fetchedAt: number;
  ttl?: number;
}

export enum SecretSource {
  ENV = 'env',
  GITHUB = 'github-actions-env',
  CLOUDFLARE = 'cloudflare',
  EXTERNAL = 'external-vault',
  CACHE = 'cache'
}

export class SecretNotFoundError extends Error {
  constructor(public key: string) {
    super(`Secret '${key}' not found in any provider`);
    this.name = 'SecretNotFoundError';
  }
}

export class SecretExpiredError extends Error {
  constructor(public key: string) {
    super(`Secret '${key}' has expired`);
    this.name = 'SecretExpiredError';
  }
}

interface SecretProvider {
  get(key: SecretKey): Promise<SecretResult | null>;
  name: string;
}

class EnvProvider implements SecretProvider {
  name = 'EnvProvider';
  
  async get(key: string): Promise<SecretResult | null> {
    const v = process.env[key];
    if (!v) return null;
    return { 
      key, 
      value: v, 
      source: SecretSource.ENV, 
      fetchedAt: Date.now() 
    };
  }
}

// GitHub Actions secrets are injected as environment variables
class GitHubActionsProvider implements SecretProvider {
  name = 'GitHubActionsProvider';
  
  async get(key: string): Promise<SecretResult | null> {
    // Check for GitHub Actions specific environment
    if (!process.env.GITHUB_ACTIONS) return null;
    
    const v = process.env[key];
    if (!v) return null;
    
    return { 
      key, 
      value: v, 
      source: SecretSource.GITHUB, 
      fetchedAt: Date.now() 
    };
  }
}

// Placeholder provider for Cloudflare (Workers/Pages binding simulation)
class CloudflareProvider implements SecretProvider {
  name = 'CloudflareProvider';
  
  constructor(private bindings?: Record<string, string>) {}
  
  async get(key: string): Promise<SecretResult | null> {
    const v = this.bindings?.[key];
    if (!v) return null;
    
    return { 
      key, 
      value: v, 
      source: SecretSource.CLOUDFLARE, 
      fetchedAt: Date.now() 
    };
  }
}

// External vault adapter placeholder (e.g., AWS SSM)
class ExternalVaultProvider implements SecretProvider {
  name = 'ExternalVaultProvider';
  
  async get(key: string): Promise<SecretResult | null> {
    // TODO: Implement actual SSM / Vault / GCP Secret Manager integration as needed.
    // This is a placeholder interface to avoid vendor lock-in
    return null;
  }
}

interface CacheEntry {
  cipher: string;
  iv: string;
  source: SecretSource;
  ts: number;
  ttl?: number;
}

export interface SecretMetrics {
  onResolve?(key: string, source: SecretSource, depth: number, latencyMs: number): void;
  onMiss?(key: string): void;
  onCacheHit?(key: string, latencyMs: number): void;
  onCacheMiss?(key: string): void;
}

export interface UnifiedSecretOptions {
  cacheTTLms?: number;
  encryptionKey?: string; // 32 bytes hex string
  metrics?: SecretMetrics;
  providers?: SecretProvider[];
}

export class UnifiedSecretResolver {
  private providers: SecretProvider[] = [];
  private cache = new Map<string, CacheEntry>();
  private encKey: Buffer;
  
  constructor(private opts: UnifiedSecretOptions = {}) {
    // Initialize encryption key
    const keySource = opts.encryptionKey || 
                     process.env.SECRET_ENCRYPTION_KEY || 
                     'royal-equips-default-dev-key-change-in-prod';
    
    if (keySource === 'royal-equips-default-dev-key-change-in-prod') {
      console.warn(JSON.stringify({
        level: 'warn',
        event: 'secret_encryption_key_default',
        message: 'Using default encryption key - set SECRET_ENCRYPTION_KEY in production'
      }));
    }
    
    this.encKey = Buffer.from(
      crypto.createHash('sha256').update(keySource).digest('hex').slice(0, 32),
      'hex'
    );
    
    // Initialize providers
    this.providers = opts.providers || [
      new EnvProvider(),
      new GitHubActionsProvider(),
      new CloudflareProvider(),
      new ExternalVaultProvider()
    ];
  }

  registerProvider(provider: SecretProvider, priority = this.providers.length) {
    this.providers.splice(priority, 0, provider);
  }

  private encrypt(plain: string): { cipher: string; iv: string } {
    const iv = crypto.randomBytes(12);
    const cipher = crypto.createCipheriv('aes-256-gcm', this.encKey, iv);
    const enc = Buffer.concat([cipher.update(plain, 'utf8'), cipher.final()]);
    const tag = cipher.getAuthTag();
    
    return { 
      cipher: Buffer.concat([enc, tag]).toString('base64'), 
      iv: iv.toString('base64') 
    };
  }

  private decrypt(entry: CacheEntry): string {
    const raw = Buffer.from(entry.cipher, 'base64');
    const tag = raw.slice(raw.length - 16);
    const data = raw.slice(0, raw.length - 16);
    const decipher = crypto.createDecipheriv(
      'aes-256-gcm',
      this.encKey,
      Buffer.from(entry.iv, 'base64')
    );
    decipher.setAuthTag(tag);
    const dec = Buffer.concat([decipher.update(data), decipher.final()]);
    return dec.toString('utf8');
  }

  private isExpired(entry: CacheEntry): boolean {
    if (!entry.ttl) return false;
    return Date.now() - entry.ts > entry.ttl;
  }

  private createKeyHash(key: string): string {
    return crypto.createHash('sha256').update(key).digest('hex').slice(0, 8);
  }

  async getSecret(key: SecretKey, explicitTTLms?: number): Promise<SecretResult> {
    const cached = this.cache.get(key);
    
    // Check cache first
    if (cached && !this.isExpired(cached)) {
      const start = performance.now();
      const value = this.decrypt(cached);
      const latencyMs = performance.now() - start;
      
      this.opts.metrics?.onCacheHit?.(this.createKeyHash(key), latencyMs);
      this.opts.metrics?.onResolve?.(this.createKeyHash(key), SecretSource.CACHE, 0, latencyMs);
      
      return {
        key,
        value,
        source: SecretSource.CACHE,
        fetchedAt: cached.ts,
        ttl: cached.ttl
      };
    }

    if (cached) {
      // Remove expired entry
      this.cache.delete(key);
    }

    this.opts.metrics?.onCacheMiss?.(this.createKeyHash(key));

    const start = performance.now();
    
    // Try each provider in order
    for (let i = 0; i < this.providers.length; i++) {
      const provider = this.providers[i];
      const res = await provider.get(key);
      
      if (res) {
        const ttl = explicitTTLms ?? this.opts.cacheTTLms ?? 5 * 60_000; // 5 minutes default
        const { cipher, iv } = this.encrypt(res.value);
        
        // Cache the encrypted result
        this.cache.set(key, {
          cipher,
          iv,
          source: res.source,
          ts: Date.now(),
          ttl
        });

        const latencyMs = performance.now() - start;
        this.opts.metrics?.onResolve?.(this.createKeyHash(key), res.source, i + 1, latencyMs);
        
        return { ...res, ttl };
      }
    }

    const keyHash = this.createKeyHash(key);
    this.opts.metrics?.onMiss?.(keyHash);
    throw new SecretNotFoundError(key);
  }

  // Get secret with fallback for graceful degradation
  async getSecretWithFallback(key: SecretKey, fallback?: string): Promise<string> {
    try {
      const result = await this.getSecret(key);
      return result.value;
    } catch (error) {
      if (error instanceof SecretNotFoundError && fallback !== undefined) {
        return fallback;
      }
      throw error;
    }
  }

  // Clear cache for testing/debugging
  clearCache(): void {
    this.cache.clear();
  }

  // Get cache stats for monitoring
  getCacheStats(): { size: number; entries: Array<{ key: string; source: string; age: number }> } {
    const entries = Array.from(this.cache.entries()).map(([key, entry]) => ({
      key: this.createKeyHash(key),
      source: entry.source,
      age: Date.now() - entry.ts
    }));

    return {
      size: this.cache.size,
      entries
    };
  }
}

// Create a default instance
export const secrets = new UnifiedSecretResolver({
  metrics: {
    onResolve: (keyHash, source, depth, latencyMs) =>
      console.log(JSON.stringify({ 
        level: 'info', 
        event: 'secret_resolve', 
        key_hash: keyHash, 
        source, 
        depth, 
        latency_ms: latencyMs,
        cache: source === SecretSource.CACHE
      })),
    onMiss: (keyHash) =>
      console.warn(JSON.stringify({ 
        level: 'warn', 
        event: 'secret_miss', 
        key_hash: keyHash 
      })),
    onCacheHit: (keyHash, latencyMs) =>
      console.log(JSON.stringify({
        level: 'debug',
        event: 'secret_cache_hit',
        key_hash: keyHash,
        latency_ms: latencyMs
      })),
    onCacheMiss: (keyHash) =>
      console.log(JSON.stringify({
        level: 'debug',
        event: 'secret_cache_miss',
        key_hash: keyHash
      }))
  }
});

File: core/secrets/secret_provider.py
"""
Multi-layer secret resolution system for Royal Equips Orchestrator.

Provides encrypted caching and fallback resolution across multiple secret sources:
ENV → GitHub Actions → Cloudflare → External Vault → Cache
"""

from __future__ import annotations
import os
import time
import base64
import secrets
import hashlib
import asyncio
from typing import Optional, Dict, List, Protocol, Any, Callable
from dataclasses import dataclass
from enum import Enum
import json
import warnings

try:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
except ImportError:
    # Fallback if cryptography not available
    AESGCM = None
    warnings.warn("cryptography package not available - encryption disabled")


class SecretSource(Enum):
    ENV = "env"
    GITHUB = "github-actions-env"
    CLOUDFLARE = "cloudflare"
    EXTERNAL = "external-vault"
    CACHE = "cache"


class SecretNotFoundError(Exception):
    """Secret not found in any provider."""
    def __init__(self, key: str):
        super().__init__(f"Secret '{key}' not found in any provider")
        self.key = key


class SecretExpiredError(Exception):
    """Secret has expired in cache."""
    def __init__(self, key: str):
        super().__init__(f"Secret '{key}' has expired")
        self.key = key


@dataclass
class SecretResult:
    key: str
    value: str
    source: SecretSource
    fetched_at: float
    ttl: Optional[int] = None


class SecretProvider(Protocol):
    """Protocol for secret providers."""
    name: str
    
    async def get(self, key: str) -> Optional[SecretResult]:
        """Get secret value for key, return None if not found."""
        ...


class EnvProvider:
    """Environment variable provider."""
    name = "EnvProvider"
    
    async def get(self, key: str) -> Optional[SecretResult]:
        v = os.getenv(key)
        if not v:
            return None
        return SecretResult(
            key=key, 
            value=v, 
            source=SecretSource.ENV, 
            fetched_at=time.time()
        )


class GitHubActionsProvider:
    """GitHub Actions secrets provider (reads from environment)."""
    name = "GitHubActionsProvider"
    
    async def get(self, key: str) -> Optional[SecretResult]:
        # Only active in GitHub Actions environment
        if not os.getenv("GITHUB_ACTIONS"):
            return None
            
        v = os.getenv(key)
        if not v:
            return None
            
        return SecretResult(
            key=key,
            value=v,
            source=SecretSource.GITHUB,
            fetched_at=time.time()
        )


class CloudflareProvider:
    """Cloudflare Workers/Pages environment provider."""
    name = "CloudflareProvider"
    
    def __init__(self, bindings: Optional[Dict[str, str]] = None):
        self.bindings = bindings or {}
    
    async def get(self, key: str) -> Optional[SecretResult]:
        v = self.bindings.get(key)
        if not v:
            return None
        return SecretResult(
            key=key,
            value=v,
            source=SecretSource.CLOUDFLARE,
            fetched_at=time.time()
        )


class ExternalVaultProvider:
    """External vault provider (AWS SSM, HashiCorp Vault, etc.)."""
    name = "ExternalVaultProvider"
    
    async def get(self, key: str) -> Optional[SecretResult]:
        # TODO: Implement actual SSM / Vault / GCP Secret Manager integration
        # This is a placeholder interface to avoid vendor lock-in
        return None


@dataclass
class SecretMetrics:
    """Metrics callbacks for secret resolution."""
    on_resolve: Optional[Callable[[str, str, int, float], None]] = None
    on_miss: Optional[Callable[[str], None]] = None
    on_cache_hit: Optional[Callable[[str, float], None]] = None
    on_cache_miss: Optional[Callable[[str], None]] = None


class UnifiedSecretResolver:
    """
    Unified secret resolver with multi-layer fallback and encrypted caching.
    
    Resolution order: ENV → GitHub Actions → Cloudflare → External Vault → Cache
    """
    
    def __init__(
        self,
        providers: Optional[List[SecretProvider]] = None,
        cache_ttl: int = 300,  # 5 minutes default
        encryption_key: Optional[bytes] = None,
        metrics: Optional[SecretMetrics] = None
    ):
        self.providers = providers or [
            EnvProvider(),
            GitHubActionsProvider(),
            CloudflareProvider(),
            ExternalVaultProvider()
        ]
        self.cache_ttl = cache_ttl
        self.metrics = metrics
        self.cache: Dict[str, Dict] = {}
        self.key = encryption_key or self._derive_key()

    def _derive_key(self) -> bytes:
        """Derive encryption key from environment or use default."""
        seed = os.getenv("SECRET_ENCRYPTION_KEY") or "royal-equips-default-dev-key-change-in-prod"
        
        if seed == "royal-equips-default-dev-key-change-in-prod":
            print(json.dumps({
                "level": "warn",
                "event": "secret_encryption_key_default",
                "message": "Using default encryption key - set SECRET_ENCRYPTION_KEY in production"
            }))
        
        # Use SHA-256 to derive 32-byte key
        return hashlib.sha256(seed.encode()).digest()

    def _encrypt(self, plaintext: str) -> Dict[str, str]:
        """Encrypt plaintext using AES-256-GCM."""
        if not AESGCM:
            # Fallback: base64 encoding (not secure, for testing only)
            return {
                "nonce": base64.b64encode(b"no-crypto-fallback").decode(),
                "cipher": base64.b64encode(plaintext.encode()).decode()
            }
        
        aes = AESGCM(self.key)
        nonce = secrets.token_bytes(12)
        ct = aes.encrypt(nonce, plaintext.encode(), None)
        
        return {
            "nonce": base64.b64encode(nonce).decode(),
            "cipher": base64.b64encode(ct).decode()
        }

    def _decrypt(self, enc: Dict[str, str]) -> str:
        """Decrypt ciphertext using AES-256-GCM."""
        if not AESGCM:
            # Fallback: base64 decoding (not secure, for testing only)
            return base64.b64decode(enc["cipher"]).decode()
        
        aes = AESGCM(self.key)
        nonce = base64.b64decode(enc["nonce"])
        ct = base64.b64decode(enc["cipher"])
        pt = aes.decrypt(nonce, ct, None)
        return pt.decode()

    def _expired(self, entry: Dict) -> bool:
        """Check if cache entry is expired."""
        ttl = entry.get("ttl")
        if ttl is None:
            return False
        return (time.time() - entry["ts"]) > ttl

    def _create_key_hash(self, key: str) -> str:
        """Create hash of key for logging (never log actual key)."""
        return hashlib.sha256(key.encode()).hexdigest()[:8]

    async def get_secret(self, key: str, ttl: Optional[int] = None) -> SecretResult:
        """
        Get secret with caching and fallback.
        
        Args:
            key: Secret key to resolve
            ttl: Cache TTL override in seconds
            
        Returns:
            SecretResult with value and metadata
            
        Raises:
            SecretNotFoundError: If secret not found in any provider
        """
        cached = self.cache.get(key)
        
        # Check cache first
        if cached and not self._expired(cached):
            start = time.time()
            value = self._decrypt(cached["data"])
            latency_ms = (time.time() - start) * 1000
            
            key_hash = self._create_key_hash(key)
            if self.metrics:
                if self.metrics.on_cache_hit:
                    self.metrics.on_cache_hit(key_hash, latency_ms)
                if self.metrics.on_resolve:
                    self.metrics.on_resolve(key_hash, SecretSource.CACHE.value, 0, latency_ms)
            
            return SecretResult(
                key=key,
                value=value,
                source=SecretSource.CACHE,
                fetched_at=cached["ts"],
                ttl=cached["ttl"]
            )

        # Remove expired entry
        if cached:
            del self.cache[key]

        key_hash = self._create_key_hash(key)
        if self.metrics and self.metrics.on_cache_miss:
            self.metrics.on_cache_miss(key_hash)

        start = time.time()
        
        # Try each provider in order
        for depth, provider in enumerate(self.providers, start=1):
            res = await provider.get(key)
            if res:
                effective_ttl = ttl or self.cache_ttl
                enc = self._encrypt(res.value)
                
                # Cache encrypted result
                self.cache[key] = {
                    "data": enc,
                    "ttl": effective_ttl,
                    "source": res.source.value,
                    "ts": time.time()
                }
                
                latency_ms = (time.time() - start) * 1000
                if self.metrics and self.metrics.on_resolve:
                    self.metrics.on_resolve(key_hash, res.source.value, depth, latency_ms)
                
                return res

        # Not found in any provider
        if self.metrics and self.metrics.on_miss:
            self.metrics.on_miss(key_hash)
        raise SecretNotFoundError(key)

    async def get_secret_with_fallback(self, key: str, fallback: Optional[str] = None) -> str:
        """Get secret with optional fallback value for graceful degradation."""
        try:
            result = await self.get_secret(key)
            return result.value
        except SecretNotFoundError:
            if fallback is not None:
                return fallback
            raise

    def clear_cache(self) -> None:
        """Clear all cached secrets."""
        self.cache.clear()

    def get_cache_stats(self) -> Dict[str, Any]:
        """Get cache statistics for monitoring."""
        entries = []
        for key, entry in self.cache.items():
            entries.append({
                "key": self._create_key_hash(key),
                "source": entry["source"],
                "age": time.time() - entry["ts"]
            })

        return {
            "size": len(self.cache),
            "entries": entries
        }

    def register_provider(self, provider: SecretProvider, priority: int = None) -> None:
        """Register a new secret provider."""
        if priority is None:
            self.providers.append(provider)
        else:
            self.providers.insert(priority, provider)


# Default metrics that log to stdout
def _default_on_resolve(key_hash: str, source: str, depth: int, latency_ms: float) -> None:
    print(json.dumps({
        "level": "info",
        "event": "secret_resolve",
        "key_hash": key_hash,
        "source": source,
        "depth": depth,
        "latency_ms": latency_ms,
        "cache": source == SecretSource.CACHE.value
    }))


def _default_on_miss(key_hash: str) -> None:
    print(json.dumps({
        "level": "warn",
        "event": "secret_miss",
        "key_hash": key_hash
    }))


def _default_on_cache_hit(key_hash: str, latency_ms: float) -> None:
    print(json.dumps({
        "level": "debug",
        "event": "secret_cache_hit",
        "key_hash": key_hash,
        "latency_ms": latency_ms
    }))


def _default_on_cache_miss(key_hash: str) -> None:
    print(json.dumps({
        "level": "debug",
        "event": "secret_cache_miss",
        "key_hash": key_hash
    }))


# Create default instance
secrets = UnifiedSecretResolver(
    metrics=SecretMetrics(
        on_resolve=_default_on_resolve,
        on_miss=_default_on_miss,
        on_cache_hit=_default_on_cache_hit,
        on_cache_miss=_default_on_cache_miss
    )
)


# Example usage
async def example():
    """Example usage of the secret resolver."""
    try:
        # Get a secret with fallback
        stripe_key = await secrets.get_secret_with_fallback("STRIPE_API_KEY", "test_key")
        print(f"Stripe key resolved (first 8 chars): {stripe_key[:8]}...")
        
        # Get cache stats
        stats = secrets.get_cache_stats()
        print(f"Cache stats: {stats}")
        
    except SecretNotFoundError as e:
        print(f"Secret not found: {e}")


if __name__ == "__main__":
    asyncio.run(example())